adam_core/observations/ades.py

```diff
@@ -0,0 +1,291 @@
+                    [
+                        ("provID", "ascending"),
+                        ("permID", "ascending"),
+                        ("trkSub", "ascending"),
+                        ("obsTime.days", "ascending"),
+                        ("obsTime.nanos", "ascending"),
+                    ]
+                )
+
+                id_present = False
+                if not pc.all(pc.is_null(observations_obscode.permID)).as_py():
+                    id_present = True
+                if not pc.all(pc.is_null(observations_obscode.provID)).as_py():
+                    id_present = True
+                if not pc.all(pc.is_null(observations_obscode.trkSub)).as_py():
+                    id_present = True
+
+                if not id_present:
+                    err = (
+                        "At least one of permID, provID, or trkSub should\n"
+                        "be present in observations."
+                    )
+                    raise ValueError(err)
+
+                # Write the observatory context block
+                obs_context = obs_contexts[obs]
+                f.writelines([obs_context.to_string()])
+
+                # Write the observations block
+                ades = observations_obscode.to_dataframe()
+
+                # Convert the timestamp to ISOT with the desired precision
+                observation_times = Time(
+                    observations_obscode.obsTime.rescale("utc")
+                    .mjd()
+                    .to_numpy(zero_copy_only=False),
+                    format="mjd",
+                    precision=seconds_precision,
+                )
+                ades.insert(
+                    4,
+                    "obsTime",
+                    np.array([i + "Z" for i in observation_times.utc.isot]),
+                )
+                ades.drop(columns=["obsTime.days", "obsTime.nanos"], inplace=True)
+
+                # Multiply rmsRA by cos(dec) since ADES wants the random component in rmsRAcosDec
+                ades.loc[:, "rmsRA"] *= np.cos(np.radians(ades["dec"]))
+
+                # Convert rmsRA and rmsDec to arcseconds
+                ades.loc[:, "rmsRA"] *= 3600
+                ades.loc[:, "rmsDec"] *= 3600
+
+                ades.dropna(how="all", axis=1, inplace=True)
+
+                # Change the precision of some of the columns to conform
+                # to MPC standards
+                for col, prec_col in columns_precision.items():
+                    print(col)
+                    if col in ades.columns:
+                        ades[col] = [
+                            f"{i:.{prec_col}f}"
+                            if i is not None or not np.isnan(i)
+                            else ""
+                            for i in ades[col]
+                        ]
+
+                # This will append to the file since the file is opened in append mode
+                ades.to_csv(f, sep="|", header=True, index=False, float_format="%.16f")
```

Comment:
Any particular reason to write to a file directly rather than to a string? Might make it easier to not have to use file objects for testing and such.

Reply:
Hmm, good point! I think I'd need to write a method to convert the dataframe to a string with the appropriate separator included. It doesn't appear that df.to_string() allows you to define the separator. I'd definitely prefer just working with a string and then having separate function write out a file with that string.